fedora_messaging/config.py

```diff
@@ -393,9 +356,56 @@ def setup_logging(self):
             self.load_config()
         logging.config.dictConfig(self['log_config'])
 
-    def load_config(self, filename=None):
+    def _validate(self):
+        """
+        Perform checks on the configuration to assert its validity
+
+        Raises:
+            ConfigurationException: If the configuration is invalid.
+        """
+        for key in self:
+            if key not in DEFAULTS:
+                raise exceptions.ConfigurationException(
+                    'Unknown configuration key "{}"! Valid configuration keys are'
+                    ' {}'.format(key, list(DEFAULTS.keys())))
+
+        # Don't let users override these
+        self['client_properties']['version'] = DEFAULTS['client_properties']['version']
+        self['client_properties']['information'] = DEFAULTS['client_properties']['information']
+        self['client_properties']['product'] = DEFAULTS['client_properties']['product']
+
+    def load_config(self):
+        """
+        Load application configuration from a file and merge it with the default
+        configuration.
+
+        If the ``FEDORA_MESSAGING_CONF`` environment variable is set to a
+        filesystem path, the configuration will be loaded from that location.
+        Otherwise, the path defaults to ``/etc/fedora-messaging/config.toml``.
+        """
         self.loaded = True
-        self.update(load(filename=filename))
+        config = DEFAULTS.copy()
+
+        if 'FEDORA_MESSAGING_CONF' in os.environ:
+            config_path = os.environ['FEDORA_MESSAGING_CONF']
+        else:
+            config_path = '/etc/fedora-messaging/config.toml'
+
+        if os.path.exists(config_path):
+            _log.info('Loading configuration from {}'.format(config_path))
+            with open(config_path) as fd:
+                try:
+                    file_config = pytoml.loads(fd.read())
+                    for key in file_config:
+                        config[key.lower()] = file_config[key]
```

Comment:
It might be good to be opinionated about case, but I don't feel strongly about it.

Reply:
@abompard, any opinion here? I don't have strong feelings about it, either.

Reply:
Me neither. I don't imagine we'll have keys that will only differ by case.